sw/device/silicon_creator/mask_rom/mask_rom_epmp.c

```diff
@@ -0,0 +1,74 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/mask_rom/mask_rom_epmp.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/csr.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+// Symbols defined in linker script.
+extern char _stack_start[];  // Lowest stack address.
+extern char _text_start[];   // Start of executable code.
+extern char _text_end[];     // End of executable code.
+
+void mask_rom_epmp_state_init(epmp_state_t *state) {
+  // Address space definitions.
+  //
+  // Note that the stack guard is placed at _stack_start because the stack
+  // grows downward from _stack_end.
+  const epmp_region_t rom_text = {.start = (uintptr_t)_text_start,
+                                  .end = (uintptr_t)_text_end};
+  const epmp_region_t rom = {.start = TOP_EARLGREY_ROM_CTRL_ROM_BASE_ADDR,
+                             .end = TOP_EARLGREY_ROM_CTRL_ROM_BASE_ADDR +
+                                    TOP_EARLGREY_ROM_CTRL_ROM_SIZE_BYTES};
+  const epmp_region_t eflash = {
+      .start = TOP_EARLGREY_EFLASH_BASE_ADDR,
+      .end = TOP_EARLGREY_EFLASH_BASE_ADDR + TOP_EARLGREY_EFLASH_SIZE_BYTES};
+  // TODO(#7117): generate MMIO addresses.
+  const epmp_region_t mmio = {.start = 0x40000000, .end = 0x50000000};
+  const epmp_region_t stack_guard = {.start = (uintptr_t)_stack_start,
+                                     .end = (uintptr_t)_stack_start + 4};
+  const epmp_region_t ram = {.start = TOP_EARLGREY_RAM_MAIN_BASE_ADDR,
+                             .end = TOP_EARLGREY_RAM_MAIN_BASE_ADDR +
+                                    TOP_EARLGREY_RAM_MAIN_SIZE_BYTES};
+
+  // Initialize shadow copy of ePMP register state.
+  //
+  // The actual hardware configuration is performed separately, either by reset
+  // logic or in assembly. This code must be kept in sync with any changes
+  // to the hardware configuration.
+  *state = (epmp_state_t){0};
+  epmp_state_configure_tor(state, 1, rom_text, kEpmpPermLockedReadExecute);
```

Comment:
Probably you've already considered this but WDYT about defining constants for the regions that will be used by mask rom and rom_ext in `epmp.h`? You could then add `static_assert`s to the functions with CSR writes that need to be updated if those definitions change. I know these are also coupled with `mask_rom_epmp.S`.

Reply:
Yeah it might be worth doing that for some of the regions. I don't think we can use `static_assert` though as we are only writing to CSRs in assembly (or based on the ROM_EXT image). I'll have a think and if it seems worthwhile I'll send a follow up.